**Цель работы:**

Осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных.

# Загрузка набора данных

### Описание предметной области

Вариант №12

Набор данных: movies.csv

Атрибуты:
...

### 1.Чтение файла (набора данных)

In [1]:
# импорт библиотек, чтение файла с помощью pandas
import pandas as pd

data = pd.read_csv('movies.csv', sep=';')


### 2. Обзор данных

2.1 Вывод первых 20 строк с помощью метода head.

In [2]:
# применить метод head
data.head(20)

title  release_year  score  \
0      David Attenborough: A Life on Our Planet        2020.0    9.0   
1                                     Inception        2010.0    8.8   
2                                  Forrest Gump        1994.0    8.8   
3                                    Anbe Sivam        2003.0    8.7   
4                            Bo Burnham: Inside        2021.0    8.7   
5                           Saving Private Ryan        1998.0    8.6   
6                              Django Unchained        2012.0    8.4   
7                                        Dangal        2016.0    8.4   
8                        Bo Burnham: Make Happy        2016.0    8.4   
9                         Louis C.K.: Hilarious        2010.0    8.4   
10              Dave Chappelle: Sticks & Stones        2019.0    8.4   
11                                     3 Idiots        2009.0    8.4   
12                                 Black Friday        2004.0    8.4   
13                                 Super Deluxe        2019.0    8.4   
14  Winter on Fire: Ukraine's Fight for Freedom        2015.0    8.3   
15                  Once Upon a Time in America        1984.0    8.3   
16                                  Taxi Driver        1976.0    8.3   
17                          Like Stars on Earth        2007.0    8.3   
18                            Bo Burnham: What.        2013.0    8.3   
19                            Full Metal Jacket        1987.0    8.3   

    number_of_votes  duration   main_genre main_production  
0           31180.0        83  documentary              GB  
1         2268288.0       148        scifi              GB  
2         1994599.0       142        drama              US  
3           20595.0       160       comedy              IN  
4           44074.0        87       comedy              US  
5         1346020.0       169        drama              US  
6         1472668.0       165      western              US  
7          180247.0       161       action              IN  
8           14356.0        60       comedy              US  
9           11973.0        84       comedy              US  
10          25687.0        65       comedy              US  
11         385782.0       170       comedy              IN  
12          20611.0       143        crime              IN  
13          13680.0       176     thriller              IN  
14          17710.0        98  documentary              UA  
15         342335.0       229        drama              US  
16         795222.0       113        crime              US  
17         188234.0       165        drama              IN  
18          11488.0        60       comedy              US  
19         723306.0       116        drama              GB

2.2 Оценка данных с помощью метода info.

In [14]:
# выполнит метод info
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 389 entries, 0 to 388
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            384 non-null    object 
 1   release_year     387 non-null    float64
 2   score            386 non-null    float64
 3   number_of_votes  388 non-null    float64
 4   duration         389 non-null    int64  
 5   main_genre       389 non-null    object 
 6   main_production  388 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 21.4+ KB


2.3 Оценка данных с помощью метода describe.

In [15]:
# оцените числовые столбцы с помощью describe

data.describe()

release_year       score  number_of_votes    duration
count    387.000000  386.000000     3.880000e+02  389.000000
mean    2010.976744    7.512176     1.391523e+05  123.352185
std       10.805726    0.443870     2.365279e+05   28.304800
min     1954.000000    6.900000     1.013900e+04   28.000000
25%     2007.500000    7.100000     2.055375e+04  104.000000
50%     2014.000000    7.400000     4.606950e+04  122.000000
75%     2018.000000    7.800000     1.551585e+05  139.000000
max     2022.000000    9.000000     2.268288e+06  229.000000


---

**Сделать выводы. Вы должны понимать, что означает тот или иной столбец, чтобы
 ответить на вопросы на защите.**


 ---


Выводы по обзору данных (пп. 1-4):

Набор данных содержит информацию о 389 фильмах с 7 атрибутами:

title: Название фильма (категориальный/строковый). Присутствуют пропуски (384 non-null).
release_year: Год выпуска фильма (числовой, тип float64). Присутствуют пропуски (387 non-null).
score: Оценка фильма (числовой, тип float64). Присутствуют пропуски (386 non-null).
number_of_votes: Количество голосов (числовой, тип float64). Присутствуют пропуски (388 non-null).
duration: Продолжительность фильма в минутах (числовой, тип int64). Пропусков нет.
main_genre: Основной жанр фильма (категориальный/строковый). Пропусков нет.
main_production: Основная страна производства фильма (категориальный/строковый). Присутствуют пропуски (388 non-null).
Типы данных:

Числовые: release_year, score, number_of_votes, duration.
Категориальные/Строковые: title, main_genre, main_production.
Пропуски обнаружены в столбцах title, release_year, score, number_of_votes и main_production. Количество пропусков незначительно по сравнению с общим количеством записей (389).

 2.4 Оценка названий столбцов

In [16]:
# Вывести на экран названия столбцов с помощью df.columns. Выявить проблемы с названиями, если они есть. При необходимости переименовать столбцы. Если проблемы не обнаружены также дать пояснения.

data.columns

Index(['title', 'release_year', 'score', 'number_of_votes', 'duration',
       'main_genre', 'main_production'],
      dtype='object')

In [17]:
# переименование при необходимости

### 3. Проверка пропусков

In [18]:
# Проверить данные на наличие пропусков и устранить их, если они есть (пропуски необходимо либо удалить, либо заменить каким-то значением).

print(data.isnull().sum())

title              5
release_year       2
score              3
number_of_votes    1
duration           0
main_genre         0
main_production    1
dtype: int64


In [19]:
data.dropna(inplace=True)

print("\nКоличество пропусков в каждом столбце после удаления:")
print(data.isnull().sum())


Количество пропусков в каждом столбце после удаления:
title              0
release_year       0
score              0
number_of_votes    0
duration           0
main_genre         0
main_production    0
dtype: int64



---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с пропусками, а также знать другие способы работы с пропусками, чтобы
 ответить на вопросы на защите.**


 ---

## **Стратегии обработки пропущенных значений в данных**
#### Выбранный подход к обработке пропусков:

Удаление записей с пропущенными значениями:

Ограничение: Риск потери репрезентативности выборки
Применимость: Только при незначительном количестве пропусков

#### Альтернативные подходы к обработке
Заполнение числовых признаков медианным значением:

Обоснование: Медиана менее чувствительна к экстремальным значениям по сравнению со средним
Преимущество: Сохранение статистических характеристик распределения без влияния выбросов

Заполнение категориальных признаков наиболее частым значением:

Обоснование: Наиболее часто встречающееся значение репрезентативно для категориальных переменных
Преимущество: Минимизация изменения исходного распределения категориальных данных

### 4. Проверка дубликатов

#### Проверка явных дубликатов

In [29]:
data.duplicated().sum()

np.int64(2)

In [30]:
# удалите дубликаты, если они есть
data.drop_duplicates(inplace=True)

# Проверим еще раз количество явных дубликатов после удаления
print("Количество явных дубликатов после удаления:", data.duplicated().sum())

Количество явных дубликатов после удаления: 0


#### Проверка неявных дубликатов

In [31]:
# Проверка на неявные дубликаты в категориальных столбцах
print("Уникальные значения в столбце 'title':")
print(data['title'].unique()[:20]) # Выводим первые 20 уникальных значений для обзора

print("\nУникальные значения в столбце 'main_genre':")
print(data['main_genre'].unique())

print("\nУникальные значения в столбце 'main_production':")
print(data['main_production'].unique())

Уникальные значения в столбце 'title':
['David Attenborough: A Life on Our Planet' 'Inception' 'Forrest Gump'
 'Anbe Sivam' 'Bo Burnham: Inside' 'Saving Private Ryan'
 'Django Unchained' 'Dangal' 'Bo Burnham: Make Happy'
 'Louis C.K.: Hilarious' 'Dave Chappelle: Sticks & Stones' '3 Idiots'
 'Black Friday' 'Super Deluxe'
 "Winter on Fire: Ukraine's Fight for Freedom"
 'Once Upon a Time in America' 'Taxi Driver' 'Like Stars on Earth'
 'Bo Burnham: What.' 'Full Metal Jacket']

Уникальные значения в столбце 'main_genre':
['documentary' 'scifi' 'drama' 'comedy' 'western' 'action' 'crime'
 'thriller' 'war' 'fantasy' 'romance' 'horror' 'animation' 'sports'
 'COMEDY' 'dramaa' 'musical']

Уникальные значения в столбце 'main_production':
['GB' 'US' 'IN' 'UA' 'CD' 'TR' 'ES' 'AU' 'JP' 'ZA' 'HK' 'DE' 'KR' 'CA'
 'BE' 'NO' 'NZ' 'MX' 'FR' 'MW' 'TH' 'AR' 'PS' 'HU' 'IT' 'CN' 'PL' 'KH'
 'IE' 'BR' 'XX' 'LT' 'NL' 'DK' 'ID']


In [33]:
# удалите дубликаты, если они есть
data['main_genre'] = data['main_genre'].replace({'COMEDY': 'comedy', 'dramaa': 'drama'})

# Проверим уникальные значения еще раз
print("\nУникальные значения в столбце 'main_genre' после исправления:")
print(data['main_genre'].unique())


Уникальные значения в столбце 'main_genre' после исправления:
['documentary' 'scifi' 'drama' 'comedy' 'western' 'action' 'crime'
 'thriller' 'war' 'fantasy' 'romance' 'horror' 'animation' 'sports'
 'musical']


---

**Вы должны аргументировать на защите, почему были выполнены те или иные действия с дубликатами, знать, что такое явные и неявные дубликаты и способы работы с ними, чтобы ответить на вопросы на защите.**


 ---

Явные дубликаты — это строки, которые полностью повторяются в наборе данных. Мы их удалили с помощью метода drop_duplicates(), потому что они представляют собой идентичные записи и могут исказить результаты анализа. Неявные дубликаты — это записи, которые по сути одинаковы, но имеют небольшие отличия (например, "Comedy" и "COMEDY"). В нашем случае, при проверке уникальных значений в категориальных столбцах, мы выявили такие неявные дубликаты в столбце main_genre ('COMEDY' и 'dramaa' вместо 'drama').

Чтобы устранить неявные дубликаты, можно привести все строковые значения к единому регистру (например, нижнему), удалить лишние пробелы или использовать метод replace() для замены некорректных значений на правильные ('COMEDY' на 'comedy', 'dramaa' на 'drama'). В данном случае, мы обнаружили 'COMEDY' и 'dramaa' в столбце main_genre. Я могу предложить код для их исправления.


### 5. Провека типов данных

In [35]:
data

title  release_year  score  \
0    David Attenborough: A Life on Our Planet        2020.0    9.0   
1                                   Inception        2010.0    8.8   
2                                Forrest Gump        1994.0    8.8   
3                                  Anbe Sivam        2003.0    8.7   
4                          Bo Burnham: Inside        2021.0    8.7   
..                                        ...           ...    ...   
382                                  The Trip        2021.0    6.9   
383                                  The Dirt        2019.0    6.9   
384                                   Top Gun        1986.0    6.9   
385                               Radhe Shyam        2022.0    6.9   
386                       Sorry to Bother You        2018.0    6.9   

     number_of_votes  duration   main_genre main_production  
0            31180.0        83  documentary              GB  
1          2268288.0       148        scifi              GB  
2          1994599.0       142        drama              US  
3            20595.0       160       comedy              IN  
4            44074.0        87       comedy              US  
..               ...       ...          ...             ...  
382          19706.0       113       comedy              NO  
383          47603.0       108        drama              US  
384         329656.0       110        drama              US  
385          21328.0       138      romance              IN  
386          75653.0       111      fantasy              US  

[381 rows x 7 columns]

In [36]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 381 entries, 0 to 386
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            381 non-null    object 
 1   release_year     381 non-null    float64
 2   score            381 non-null    float64
 3   number_of_votes  381 non-null    float64
 4   duration         381 non-null    int64  
 5   main_genre       381 non-null    object 
 6   main_production  381 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 23.8+ KB


In [37]:
data['release_year'] = data['release_year'].astype(int)
data['number_of_votes'] = data['number_of_votes'].astype(int)

print(data.info())

<class 'pandas.core.frame.DataFrame'>
Index: 381 entries, 0 to 386
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            381 non-null    object 
 1   release_year     381 non-null    int64  
 2   score            381 non-null    float64
 3   number_of_votes  381 non-null    int64  
 4   duration         381 non-null    int64  
 5   main_genre       381 non-null    object 
 6   main_production  381 non-null    object 
dtypes: float64(1), int64(3), object(3)
memory usage: 23.8+ KB
None


In [22]:
# Проверьте типы данных, при необходимости измените типы данных, чтобы они соответствовали действительности.

---

**Обратите внимание, что во всех вариантах необходимо сделать приведение типов. Будьте готовы на защите аргументировать проверку типов (почему выполнены те или иные преобразования).**


 ---

Обоснование выбора типов: release_year, number_of_votes - целочисленные значения

### 6. Группировка данных

#### Задание 1

Группировка - год выхода и количество фильмов каждого жанра
(main_genre).

In [41]:
genre_counts_by_year = data.groupby(['release_year', 'main_genre']).size().reset_index(name='number_of_movies')

print("Количество фильмов по году выхода и жанру:")
print(genre_counts_by_year)


Количество фильмов по году выхода и жанру:
     release_year main_genre  number_of_movies
0            1954    romance                 1
1            1961        war                 1
2            1964      drama                 1
3            1966    western                 1
4            1967      drama                 1
..            ...        ...               ...
159          2021   thriller                 2
160          2022     comedy                 1
161          2022      crime                 1
162          2022      drama                 2
163          2022    romance                 1

[164 rows x 3 columns]


*`Формулировка задания`*

In [23]:
# выполните группировку согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

На основе результатов группировки по году выхода и жанру (genre_counts_by_year) можно увидеть распределение фильмов по жанрам в разные годы. Например, можно определить:

 - Какие жанры были популярны в определенные годы.
 - Как менялось количество фильмов определенного жанра с течением времени.
 - В какие годы выходило больше всего фильмов в целом.

Этот результат помогает понять тенденции в кинематографе с точки зрения жанров и их динамики по годам.

---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 2

*`Формулировка задания`*

Группировка - год выхода и количество фильмов каждого жанра
(main_genre). Создать датафрейм. Переименовать столбец с количеством в
“сount”. Отсортировать по убыванию столбца “count”.

In [43]:
genre_counts_by_year = data.groupby(['release_year', 'main_genre']).size().reset_index(name='count')

genre_counts_by_year_sorted = genre_counts_by_year.sort_values(by='count', ascending=False)

print("Количество фильмов по году выхода и жанру (отсортировано по количеству):")
print(genre_counts_by_year_sorted)

Количество фильмов по году выхода и жанру (отсортировано по количеству):
     release_year main_genre  count
121          2017      drama     13
139          2019      drama     13
147          2020      drama      9
103          2015      drama      9
155          2021      drama      9
..            ...        ...    ...
157          2021      scifi      1
158          2021     sports      1
160          2022     comedy      1
161          2022      crime      1
163          2022    romance      1

[164 rows x 3 columns]


In [24]:
# выполните группировку согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной группировки?`**

Результат этой группировки (genre_counts_by_year_sorted), отсортированный по убыванию количества фильмов, позволяет быстро увидеть наиболее "продуктивные" сочетания года выпуска и жанра. Можно определить:

 - В какие годы и в каких жанрах было выпущено наибольшее количество фильмов.
 - Наиболее популярные жанры в определенные периоды времени с точки зрения объема производства.

Например, верхние строки отсортированной таблицы показывают, что в 2017 и 2019 годах жанр "drama" был наиболее представлен с 13 фильмами в каждом году.



---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 3

*`Формулировка задания`*

Сводная таблица (pivot_table) - средний рейтинг (score) фильмов по
жанрам. Отсортировать по убыванию рейтинга. Округлить до двух знаков.

In [45]:
genre_score_pivot = data.pivot_table(index='main_genre', values='score', aggfunc='mean')

genre_score_pivot_sorted = genre_score_pivot.sort_values(by='score', ascending=False)

genre_score_pivot_sorted['score'] = genre_score_pivot_sorted['score'].round(2)

print("Средний рейтинг фильмов по жанрам (отсортировано по убыванию рейтинга):")
print(genre_score_pivot_sorted)

Средний рейтинг фильмов по жанрам (отсортировано по убыванию рейтинга):
             score
main_genre        
war           7.80
scifi         7.77
documentary   7.76
action        7.70
comedy        7.61
western       7.56
fantasy       7.48
drama         7.48
crime         7.47
thriller      7.47
romance       7.40
sports        7.40
horror        7.35
animation     7.20
musical       7.00


In [25]:
# выполните сводную таблицу согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата данной сводной таблицы?`**

На основе результатов сводной таблицы (genre_score_pivot_sorted) можно увидеть средний рейтинг фильмов для каждого жанра, отсортированный от самого высокого к самому низкому. Это позволяет определить:

 - Какие жанры имеют наивысший средний рейтинг среди зрителей (например, "war" и "scifi" в данном случае).
 - Какие жанры имеют более низкий средний рейтинг (например, "animation" и "musical").
 - Сравнить средние рейтинги разных жанров между собой.




---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

#### Задание 4

*`Формулировка задания`*

Сводная таблица (pivot_table) - средний рейтинг (score) фильмов по
жанрам - строки и годам - столбцы. Отсортировать по возрастанию main_genre.


In [47]:
genre_year_score_pivot = data.pivot_table(index='main_genre', columns='release_year', values='score', aggfunc='mean')

genre_year_score_pivot_sorted = genre_year_score_pivot.sort_index(ascending=True)

print("Средний рейтинг фильмов по жанрам (строки) и годам (столбцы):")
print(genre_year_score_pivot_sorted)

Средний рейтинг фильмов по жанрам (строки) и годам (столбцы):
release_year  1954  1961  1964  1966  1967  1971  1973  1975  1976  1979  ...  \
main_genre                                                                ...   
action         NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
animation      NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
comedy         NaN   NaN   NaN   NaN   NaN   NaN   NaN   8.2   NaN   7.8  ...   
crime          NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   8.3   NaN  ...   
documentary    NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
drama          NaN   NaN   7.8   NaN   7.7   NaN   NaN   NaN   NaN   NaN  ...   
fantasy        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
horror         NaN   NaN   NaN   NaN   NaN   NaN   8.1   NaN   NaN   NaN  ...   
musical        NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  ...   
romance        7.5   NaN   NaN   NaN   NaN   Na

In [26]:
# выполните сводную таблицу согласно варианту

**`Сделайте вывод - интерпретацию - что можно увидеть из результата выполнения данного кода?`**

На основе результатов выполнения кода, который создает сводную таблицу средних рейтингов фильмов по жанрам (строки) и годам выпуска (столбцы), можно увидеть:

 - Динамику среднего рейтинга по жанрам с течением времени: можно проследить, как менялся средний рейтинг определенного жанра из года в год.
 - Сравнение средних рейтингов разных жанров в конкретные годы: Можно сравнить, какие жанры имели более высокие или низкие средние рейтинги в определенный год.
 - Годы с наибольшей активностью по жанрам: Ячейки с числовыми значениями указывают на наличие фильмов данного жанра в соответствующем году. Пустые ячейки (NaN) означают отсутствие данных (фильмов) для данного сочетания жанра и года в вашем наборе данных.


---

**Обратите внимание, что на защите вы должны ориентироваться в синтаксисе. При необходимости нужно быть готовым изменить код по просьбе преподавателя. Например, вместо среднего значения подсчитать медиану и т.д.**


 ---

### Вывод


***`Сделать расширенный вывод. `***

Вывод должен описывать результаты исследования: какой набор данных вы анализировали (предметная область), какие методы предобработки применили, какие выводы из группировок и сводных таблиц можно сделать (они могут повторять выводы выше) и т.п. Объём вывода, как правило не более 0.5-0.75 страницы А4, всё таки вывод должен быть конкретным (без теории и “воды”).


В рамках лабораторной работы был проанализирован набор данных movies.csv, содержащий сведения о 389 фильмах: название, год выпуска, рейтинг, количество голосов, продолжительность, жанр и страну производства.

Предобработка данных включала загрузку набора в DataFrame, анализ структуры и удаление строк с пропусками и дубликатами. Были устранены неявные дубликаты в категориальных признаках (например, некорректные значения жанров), приведены типы данных к корректным (например, release_year и number_of_votes — к целочисленным). После очистки в датасете осталось 381 запись без пропусков и дубликатов.

Группировки и сводные таблицы позволили выявить ключевые закономерности:

 - Количество фильмов по годам и жанрам показало периоды наибольшей активности кинопроизводства и доминирующие жанры в разные годы.

 - Отсортированная группировка подтвердила, что в 2017–2019 годах жанр drama был наиболее представлен.

 - Сводная таблица среднего рейтинга по жанрам показала, что наивысшие оценки получили жанры war, scifi и documentary, тогда как animation и musical имели более низкие значения.

 - Анализ динамики рейтингов по жанрам и годам позволил проследить изменения зрительских предпочтений.

В результате была проведена полная предобработка и базовый исследовательский анализ данных о фильмах. Полученные сводки и группировки дают представление о распределении жанров, динамике их популярности и различиях в оценках, что создаёт основу для дальнейшего анализа тенденций в киноиндустрии.

### Дополнительное задание

**`Подробная формулировка задания`**

In [27]:
# код выполнения задания

***`Подробный вывод по заданию, описание полученных результатов`***